plugins/transport-nio/src/main/java/org/elasticsearch/transport/nio/SharedNioGroup.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio;
+
+import org.elasticsearch.common.util.concurrent.AbstractRefCounted;
+import org.elasticsearch.nio.ChannelFactory;
+import org.elasticsearch.nio.EventHandler;
+import org.elasticsearch.nio.NioGroup;
+import org.elasticsearch.nio.NioSelector;
+import org.elasticsearch.nio.NioServerSocketChannel;
+import org.elasticsearch.nio.NioSocketChannel;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.InetSocketAddress;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class SharedNioGroup implements AutoCloseable {
+
+    private final RefCountedNioGroup refCountedNioGroup;
+
+    private final AtomicBoolean isOpen = new AtomicBoolean(true);
+
+    SharedNioGroup(ThreadFactory threadFactory, int selectorCount, Function<Supplier<NioSelector>, EventHandler> eventHandlerFunction)
+        throws IOException {
+        this(new RefCountedNioGroup(new NioGroup(threadFactory, selectorCount, eventHandlerFunction)));
+    }
+
+    private SharedNioGroup(RefCountedNioGroup refCountedNioGroup) {
+        this.refCountedNioGroup = refCountedNioGroup;
+    }
+
+    public <S extends NioServerSocketChannel> S bindServerChannel(InetSocketAddress address, ChannelFactory<S, ?> factory)
+        throws IOException {
+        ensureOpen();
+        return refCountedNioGroup.nioGroup.bindServerChannel(address, factory);
+    }
+
+    public <S extends NioSocketChannel> S openChannel(InetSocketAddress address, ChannelFactory<?, S> factory) throws IOException {
+        ensureOpen();
+        return refCountedNioGroup.nioGroup.openChannel(address, factory);
+    }
+
+    @Override
+    public void close() {
+        if (isOpen.compareAndSet(true, false)) {
+            refCountedNioGroup.decRef();
```

Comment:
You `incRef` the nioGroup everytime somebody calls `getSharedInstance` and then pass on the reference to the new `SharedNioGroup` that ref is decremented if that group is then closed. Wouldn't it make the code simpler if you would extend `AbstractRefCounted` on `SharedNioGroup` and then just use this interface and always return the same instance? I don't see what the indirection buys us here except of complexity?

Reply:
I am happy to go with your approach. The reasoning behind my original approach is that you can "close" specific shared reference. So essentially if you close the shared usage in the `http` package you can no longer open new channels using that instance, even though the underlying ref counted `NioGroup` is still open.

If we went with a single ref counted instance, you could still use it in the http package after decrementing the reference if the underlying group is still open.

Let me know which approach you want me to take.

Reply:
I think what would make it simpler is to make `NioGroup extends AbstractRefCounted` that way you can just create a new `SharedNioGroup` in `NioGroupFactory` and the creation vs. calling `getSharedInstance` would be less confusing. You still have the advantage of the closing which I see being useful but don't have this `getSharedInstance` method and `SharedNioGroup` is just a simple wrapper. Even better I think it would be if `NioGroup` would be an interface then we wouldn't need `ShardNioGroup` being public and we can just wrap it?